src/common/snippets/src/pass/transpose_decomposition.cpp

```diff
@@ -0,0 +1,115 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <snippets/pass/transpose_decomposition.hpp>
+#include <snippets/itt.hpp>
+#include <snippets/snippets_isa.hpp>
+#include <snippets/op/loop_helpers.hpp>
+
+#include <ngraph/opsets/opset1.hpp>
+#include <ngraph/opsets/opset5.hpp>
+#include <ngraph/partial_shape.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/pass/manager.hpp>
+#include <numeric>
+
+ngraph::snippets::pass::TransposeDecomposition::TransposeDecomposition() {
+    MATCHER_SCOPE(TransposeDecomposition);
+    // todo: we need a special transformation that detects and propagates data access pattern to Parameters and Results
+    //  this is needed to communicate access pattern to the plugin node and op::Kernel
+    // This is the reason we me match only to Parameter, this limitation could be relaxed if we propagate access pattern
+    // to the appropriate parameter
+    auto match_data = ngraph::pattern::wrap_type<opset1::Parameter>();
+    auto match_order = ngraph::pattern::wrap_type<opset1::Constant>();
+    auto match_transpose = ngraph::pattern::wrap_type<ngraph::opset1::Transpose>({match_data, match_order});
+
+    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher& m) {
+        OV_ITT_SCOPED_TASK(ngraph::pass::itt::domains::SnippetsTransform, "Snippets::op::TransposeDecomposition")
+        auto& pattern_to_output = m.get_pattern_value_map();
+        const auto transpose = ov::as_type_ptr<ngraph::opset1::Transpose>(
+                                                            pattern_to_output.at(match_transpose).get_node_shared_ptr());
+
+        const auto order = ov::as_type_ptr<ov::op::v0::Constant>(pattern_to_output.at(match_order).get_node_shared_ptr());
+        if (transformation_callback(transpose) || transpose->is_dynamic() ||
+            order->get_output_element_type(0) != ov::element::i32)
+            return false;
+
+        auto order_value = order->get_vector<int>();
+        std::set<std::vector<int>> supported_cases{{0, 2, 3, 1}};
```

Comment:
Maybe it should be as `MatcherPass` class field. User can see in header which transpose cases callback supports. What do you think?

Reply:
Can't agree more, thanks! Now we'll be able to check supported_cases directly from collapse_subgraph, so we'll avoid hardcoding supported order in two different places.